src/util/expr_iterator.h

```diff
@@ -0,0 +1,211 @@
+/*******************************************************************\
+
+ Module: exprt iterator module
+
+ Author: DiffBlue Limited. All rights reserved.
+
+\*******************************************************************/
+
+#ifndef CPROVER_UTIL_EXPR_ITERATOR_H
+#define CPROVER_UTIL_EXPR_ITERATOR_H
+
+#include <iterator>
+#include <functional>
+#include <set>
+#include <algorithm>
+#include "expr.h"
+#include "invariant.h"
+
+// Forward depth-first search iterators
```

Comment:
Could all the comments be turned into doxygen comments? This should become a very useful facility, thus people need to be able to understand it.